File: hash/js/rosetta-squint-hash/tests/group4.test.ts
import { describe, it, expect } from "vitest";
import { hexToHash, hexToFlathash, ImageHashError } from "../src/index.js";

describe("hex round-trip (Group 4)", () => {
  it("hex_to_hash and back", () => {
    const hex = "ffd7918181c9ffff";
    const h = hexToHash(hex);
    expect(h.bitCount()).toBe(64);
    expect(h.toHex()).toBe(hex);
  });

  it("hex_to_flathash and back", () => {
    const hex = "0123456789abcd";
    const h = hexToFlathash(hex, 4);
    expect(h.bitCount()).toBe(14 * 4);
    expect(h.toHex()).toBe(hex);
  });

  it("hex_to_hash rejects non-square length", () => {
    expect(() => hexToHash("12345")).toThrow(ImageHashError);
  });

  it("hex_to_hash rejects invalid chars", () => {
    expect(() => hexToHash("xyz!")).toThrow(ImageHashError);
  });

  it("round-trip all zeros", () => {
    const hex = "0000000000000000";
    expect(hexToHash(hex).toHex()).toBe(hex);
  });
});

File: hash/rust/rosetta-squint-hash/tests/group4_test.rs
use rosetta_squint_hash::{hex_to_flathash, hex_to_hash, ImageHashError};

#[test]
fn hex_to_hash_and_back() {
    let hex = "ffd7918181c9ffff";
    let h = hex_to_hash(hex).expect("parse");
    assert_eq!(h.bit_count(), 64);
    assert_eq!(h.to_hex(), hex);
}

#[test]
fn hex_to_flathash_and_back() {
    let hex = "0123456789abcd";
    let h = hex_to_flathash(hex, 4).expect("parse");
    assert_eq!(h.bit_count(), 14 * 4);
    assert_eq!(h.to_hex(), hex);
}

#[test]
fn hex_to_hash_non_square_errors() {
    match hex_to_hash("12345") {
        Err(ImageHashError::InvalidHex(_)) => (),
        other => panic!("expected InvalidHex, got {other:?}"),
    }
}

#[test]
fn hex_to_hash_invalid_chars_errors() {
    match hex_to_hash("xyz!") {
        Err(ImageHashError::InvalidHex(_)) => (),
        other => panic!("expected InvalidHex, got {other:?}"),
    }
}

#[test]
fn round_trip_all_zeros() {
    let hex = "0000000000000000";
    let h = hex_to_hash(hex).expect("parse");
    assert_eq!(h.to_hex(), hex);
}
